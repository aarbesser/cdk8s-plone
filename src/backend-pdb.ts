// eslint-disable-next-line import/no-extraneous-dependencies
import { Construct } from 'constructs';
import { IntOrString, KubePodDisruptionBudget } from './imports/k8s';

export interface PloneBackendPDBOptions {
  /**
   * maxUnavailable specification
   * @default - none
   */
  readonly maxUnavailable?: number | string;

  /**
   * minAvailable specification.
   * @default 1
   */
  readonly minAvailable?: number | string;

  /**
   * Selector label.
   */
  readonly selectorLabel: { [name: string]: string };

  /**
   * Extra labels to associate with resources.
   * @default - none
   */
  readonly labels?: { [name: string]: string };
}

export class PloneBackendPDB extends Construct {

  constructor(scope: Construct, id: string, options: PloneBackendPDBOptions) {
    super(scope, id);

    var maxUnavailable: IntOrString = IntOrString.fromString(''); // default value
    if (typeof maxUnavailable === 'number') {
      maxUnavailable = IntOrString.fromNumber(options.maxUnavailable as number);
    } else if (typeof maxUnavailable === 'string') {
      maxUnavailable = IntOrString.fromString(options.maxUnavailable as string);
    }
    var minAvailable: IntOrString = IntOrString.fromString(''); // default value
    if (typeof minAvailable === 'number') {
      minAvailable = IntOrString.fromNumber(options.minAvailable as number);
    } else if (typeof minAvailable === 'string') {
      minAvailable = IntOrString.fromString(options.minAvailable as string);
    }

    new KubePodDisruptionBudget(this, 'PDB', {
      metadata: {
        labels: options.labels ?? {},
      },
      spec: {
        selector: { matchLabels: options.selectorLabel },
        maxUnavailable: maxUnavailable,
        minAvailable: minAvailable,
      },
    });
  }
}
