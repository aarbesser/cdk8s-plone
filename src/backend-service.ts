// eslint-disable-next-line import/no-extraneous-dependencies
import { Construct } from 'constructs';
import { IntOrString, KubeServiceProps, KubeService } from './imports/k8s';

export interface PloneBackendServiceOptions {
  /**
   * Port number.
   * @default 8080
   */
  readonly port?: number;

  /**
   * Port number.
   * @default 8080;
   */
  readonly targetPort?: number;

  /**
   * Extra labels to associate with resources.
   * @default - none
   */
  readonly labels?: { [name: string]: string };
}

export class PloneBackendService extends Construct {

  constructor(scope: Construct, id: string, options: PloneBackendServiceOptions = {}) {
    super(scope, id);

    const port = options.port ?? 8080;
    const targetPort = IntOrString.fromNumber(options.targetPort ?? 8080);
    const labels = options.labels ?? {};
    const name = id + '-service';

    const serviceOpts: KubeServiceProps = {
      metadata: {
        name: name,
      },
      spec: {
        type: 'ClusterIP',
        clusterIp: 'None',
        ports: [{ port: port, targetPort: targetPort }],
        selector: labels,
      },
    };
    new KubeService(this, 'service', serviceOpts);
  }
}
