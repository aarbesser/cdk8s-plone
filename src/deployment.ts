import { Names } from 'cdk8s';
import * as kplus from 'cdk8s-plus-24';
import { Construct } from 'constructs';
import * as k8s from './imports/k8s';
import { PlonePDB, PlonePDBOptions } from './pdb';

export interface PloneImageOptions {
  /**
   * Specify a custom image for Plone .
   * @default ""
   */
  readonly image?: string;

  /**
   * Specify Secret to pull image .
   * @default ""
   */
  readonly imagePullSecret?: string;

  /**
   * Specify Pull Policy .
   * @default ""
   */
  readonly imagePullPolicy?: string;
}

export interface PloneDeploymentOptions {
  /**
   * Specify a custom image for Plone .
   * @default "plone/plone-backend:latest"
   */
  readonly image?: PloneImageOptions;

  /**
   * Specify an environment for Plone .
   * @default - none
  */
  readonly environment?: kplus.Env;

  /**
 * Number of replicas.
 * @default 2
 */
  readonly replicas?: number;

  /**
   * Port number.
   */
  readonly port: number;

  /**
   * Extra labels to associate with resources.
   * @default - none
   */
  readonly labels?: { [name: string]: string };

  /**
   * Extra container spec to use for ploencontainer.
   * @default - []
   */
  readonly ploneContainer?: k8s.Container;

  /**
   * Sidecar container spec to associate with resources.
   * @default - []
   */
  readonly sidecarContainers?: k8s.Container[];

  /**
   * Create a PodDisruptionBugdet for the deployment?
   * If given
   * @default - none
   */
  readonly pdb?: PlonePDBOptions;
}

export class PloneDeployment extends Construct {

  constructor(scope: Construct, id: string, options: PloneDeploymentOptions) {
    super(scope, id);
    const image = options.image ?? {};
    const replicas = options.replicas ?? 2;
    const label = { app: Names.toLabelValue(this) };
    const template_labels = {
      ...options.labels ?? {},
      ...label,
    };
    const kpEnv = options.environment ?? new kplus.Env([], {});
    var env: k8s.EnvVar[] = [];
    for (const name in kpEnv.variables) {
      env.push({ name: name, value: kpEnv.variables[name].value });
    }
    var ploneContainerSpec: k8s.Container = {
      name: id + '-container', // here the namespaced name shold be used, but how?
      image: image.image,
      imagePullPolicy: image.imagePullPolicy,
      // imagePullSecret: image.imagePullSecret, -> ServiceAccount should be used
      env: env,
    };
    const deploymentOptions: k8s.KubeDeploymentProps = {
      metadata: {
        labels: options.labels ?? {},
      },
      spec: {
        replicas,
        selector: {
          matchLabels: label,
        },
        template: {
          metadata: { labels: template_labels },
          spec: {
            containers: [
              ploneContainerSpec,
              ...options.sidecarContainers ?? [],
            ],
          },
        },
      },
    };

    new k8s.KubeDeployment(this, 'deployment', deploymentOptions);

    if (options.pdb ?? false) {
      const pdbOptions = options.pdb ?? {};
      new PlonePDB(this, 'pdb', label, pdbOptions);
    }
  }
}
