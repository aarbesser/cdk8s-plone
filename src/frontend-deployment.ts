import { Names } from 'cdk8s';
// eslint-disable-next-line import/no-extraneous-dependencies
import { Construct } from 'constructs';
import * as k8s from './imports/k8s';
import { PlonePDB, PlonePDBOptions } from './pdb';

export interface PloneFrontendDeploymentOptions {
  /**
   * Specify a custom image for Plone Frontend.
   * @default "plone/plone-frontend:latest"
   */
  readonly image?: string;

  /**
 * Number of replicas.
 * @default 2
 */
  readonly replicas?: number;

  /**
   * Port number.
   * @default 8080
   */
  readonly port?: number;

  /**
   * Extra labels to associate with resources.
   * @default - none
   */
  readonly labels?: { [name: string]: string };

  /**
   * Create a PodDisruptionBugdet for the deployment?
   * If given
   * @default - none
   */
  readonly pdbOptions?: PlonePDBOptions;
}

export class PloneFrontendDeployment extends Construct {

  constructor(scope: Construct, id: string, options: PloneFrontendDeploymentOptions = {}) {
    super(scope, id);
    const image = options.image ?? 'plone/plone-backend:latest';
    const replicas = options.replicas ?? 2;
    const label = { app: Names.toLabelValue(this) };
    const template_labels = {
      ...options.labels ?? {},
      ...label,
    };
    const deploymentOptions: k8s.KubeDeploymentProps = {
      metadata: {
        labels: options.labels ?? {},
      },
      spec: {
        replicas,
        selector: {
          matchLabels: label,
        },
        template: {
          metadata: { labels: template_labels },
          spec: {
            containers: [
              {
                name: id + '-container', // here the namespaced name shold be used, but how?
                image: image,
              },
            ],
          },
        },
      },
    };

    new k8s.KubeDeployment(this, 'deployment', deploymentOptions);


    if (options.pdbOptions ?? false) {
      const pdbOptions = options.pdbOptions ?? {};
      new PlonePDB(this, id + '-pdb', label, pdbOptions);
    }
  }
}
