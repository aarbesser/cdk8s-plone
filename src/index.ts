export { Plone, PloneOptions, PloneBaseOptions } from './plone';
