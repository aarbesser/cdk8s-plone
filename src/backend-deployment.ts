import { Names } from 'cdk8s';
// eslint-disable-next-line import/no-extraneous-dependencies
import { Construct } from 'constructs';
import { PloneBackendPDB, PloneBackendPDBOptions } from './backend-pdb';
import * as k8s from './imports/k8s';

export interface PloneBackendDeploymentOptions {
  /**
   * Specify a custom image for Plone Backend.
   * @default "plone/plone-backend:latest"
   */
  readonly image?: string;

  /**
 * Number of replicas.
 * @default 2
 */
  readonly replicas?: number;

  /**
   * Port number.
   * @default 8080
   */
  readonly port?: number;

  /**
   * Extra labels to associate with resources.
   * @default - none
   */
  readonly labels?: { [name: string]: string };

  /**
   * Create a PodDisruptionBugdet for the deployment?
   * If given
   * @default - none
   */
  readonly pdbOptions?: PloneBackendPDBOptions;
}

export class PloneBackendDeployment extends Construct {

  constructor(scope: Construct, id: string, options: PloneBackendDeploymentOptions = { }) {
    super(scope, id);
    const image = options.image ?? 'plone/plone-backend:latest';
    const replicas = options.replicas ?? 2;
    const label = { app: Names.toLabelValue(this) };
    const template_labels = {
      ...options.labels ?? {},
      ...label,
    };
    const pdb = options.pdbOptions ?? true;
    const deploymentOptions: k8s.KubeDeploymentProps = {
      metadata: {
        labels: options.labels ?? {},
      },
      spec: {
        replicas,
        selector: {
          matchLabels: label,
        },
        template: {
          metadata: { labels: template_labels },
          spec: {
            containers: [
              {
                name: id + '-container', // here the namespaced name shold be used, but how?
                image: image,
              },
            ],
          },
        },
      },
    };

    new k8s.KubeDeployment(this, 'deployment', deploymentOptions);

    if (pdb ?? false) {
      const pdbOptions = {
        ...options.pdbOptions ?? {},
        selector_label: { app: Names.toLabelValue(this) },
      };
      new PloneBackendPDB(this, 'pdb', pdbOptions);
    }
  }
}
