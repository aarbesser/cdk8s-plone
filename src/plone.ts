import { Names } from 'cdk8s';
import * as kplus from 'cdk8s-plus-24';
import { Construct } from 'constructs';
import { PloneDeployment } from './deployment';
import { PloneService } from './service';

export interface PloneOptions {
  readonly backendImage?: string;
  readonly backendImagePullSecret?: string;
  readonly backendImagePullPolicy?: string;
  readonly backendReplicas?: number;
  readonly backendMaxUnavailable?: number | string;
  readonly backendMinAvailable?: number | string;
  readonly backendEnvironment?: kplus.Env;

  readonly frontendImage?: string;
  readonly frontendImagePullSecret?: string;
  readonly frontendImagePullPolicy?: string;
  readonly frontendReplicas?: number;
  readonly frontendMaxUnavailable?: number | string;
  readonly frontendMinAvailable?: number | string;
  readonly frontendEnvironment?: kplus.Env;
}

export class Plone extends Construct {

  public readonly backendServiceName: string;
  public readonly frontendServiceName: string;

  constructor(scope: Construct, id: string, options: PloneOptions = {}) {
    super(scope, id);

    // Backend
    const backendPort = 8080;
    const backendDeployment = new PloneDeployment(this, 'backend', {
      image: {
        image: options.backendImage ?? 'plone/plone-backend:latest',
        imagePullSecret: options.backendImagePullSecret ?? '',
        imagePullPolicy: options.backendImagePullPolicy ?? 'ifNotPresent',
      },
      replicas: options.backendReplicas,
      pdb: {
        maxUnavailable: options.backendMaxUnavailable ?? undefined,
        minAvailable: options.backendMinAvailable ?? undefined,
      },
      port: backendPort,
      environment: options.backendEnvironment,
    });
    const backendService = new PloneService(backendDeployment, 'service', {
      targetPort: backendPort,
      selectorLabel: { app: Names.toLabelValue(backendDeployment) },
    });
    this.backendServiceName = backendService.name;

    // Frontend
    const frontendPort = 3000;
    var frontendEnvironment = options.frontendEnvironment ?? new kplus.Env([], {});
    if (frontendEnvironment.variables.RAZZLE_API_PATH === undefined) {
      frontendEnvironment?.addVariable('RAZZLE_INTERNAL_API_PATH', kplus.EnvValue.fromValue(`http://${backendService.name}:80`));
    }

    const frontendDeployment = new PloneDeployment(this, 'frontend', {
      image: {
        image: options.frontendImage ?? 'plone/plone-frontend:latest',
        imagePullSecret: options.frontendImagePullSecret ?? '',
        imagePullPolicy: options.frontendImagePullPolicy ?? 'ifNotPresent',
      },
      replicas: options.frontendReplicas,
      pdb: {
        maxUnavailable: options.frontendMaxUnavailable ?? undefined,
        minAvailable: options.frontendMinAvailable ?? undefined,
      },
      port: frontendPort,
      environment: frontendEnvironment,
    });
    const frontendService = new PloneService(frontendDeployment, 'service', {
      targetPort: frontendPort,
      selectorLabel: { app: Names.toLabelValue(frontendDeployment) },
    });
    this.frontendServiceName = frontendService.name;
  }
}