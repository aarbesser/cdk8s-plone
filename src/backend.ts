// eslint-disable-next-line import/no-extraneous-dependencies
import { Construct } from 'constructs';
import { PloneBackendDeploymentOptions, PloneBackendDeployment } from './backend-deployment';
import { PloneBackendServiceOptions, PloneBackendService } from './backend-service';

export interface PloneBackendOptions {
  readonly deployment?: PloneBackendDeploymentOptions;
  readonly service?: PloneBackendServiceOptions;
}

export class PloneBackend extends Construct {

  constructor(scope: Construct, id: string, options: PloneBackendOptions = {}) {
    super(scope, id);
    const deployment = options.deployment ?? {};
    const service = options.service ?? {};

    // Create a deployment
    new PloneBackendDeployment(this, 'deployment', deployment);

    // Create a service
    new PloneBackendService(this, 'service', service);
  }
}
