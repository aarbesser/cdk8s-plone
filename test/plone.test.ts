import { Chart, Testing } from 'cdk8s';
import { Plone, PloneOptions } from '../src/plone';


test('defaults', () => {
  // GIVEN
  const app = Testing.app();
  const chart = new Chart(app, 'app');

  // WHEN
  new Plone(chart, 'plone');

  // THEN
  expect(Testing.synth(chart)).toMatchSnapshot();
});

test('defaults-with-pdps', () => {
  // GIVEN
  const app = Testing.app();
  const chart = new Chart(app, 'app');
  const options: PloneOptions = {
    frontend: {
      deployment: {
        pdb: {
          minAvailable: 1,
        },
      },
    },
    backend: {
      deployment: {
        pdb: {
          minAvailable: 1,
        },
      },
    },
  };

  // WHEN
  new Plone(chart, 'plone_with_pdbs', options);

  // THEN
  expect(Testing.synth(chart)).toMatchSnapshot();
});
