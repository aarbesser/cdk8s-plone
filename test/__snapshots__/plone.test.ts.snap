// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defaults 1`] = `
[
  {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": {
      "name": "app-plone-backend-deployment-c85baacf",
    },
    "spec": {
      "replicas": 2,
      "selector": {
        "matchLabels": {
          "app": "app-plone-backend-c82290e6",
        },
      },
      "template": {
        "metadata": {
          "labels": {
            "app": "app-plone-backend-c82290e6",
          },
        },
        "spec": {
          "containers": [
            {
              "env": [],
              "image": "plone/plone-backend:latest",
              "imagePullPolicy": "ifNotPresent",
              "name": "backend-container",
            },
          ],
        },
      },
    },
  },
  {
    "apiVersion": "policy/v1",
    "kind": "PodDisruptionBudget",
    "metadata": {
      "name": "app-plone-backend-pdb-c8f00fdc",
    },
    "spec": {
      "minAvailable": 1,
      "selector": {
        "matchLabels": {
          "app": "app-plone-backend-c82290e6",
        },
      },
    },
  },
  {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": {
      "name": "app-plone-backend-service-c8efef4c",
    },
    "spec": {
      "clusterIP": "None",
      "ports": [
        {
          "port": 80,
          "targetPort": 8080,
        },
      ],
      "selector": {
        "app": "app-plone-backend-c82290e6",
      },
      "type": "ClusterIP",
    },
  },
  {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": {
      "name": "app-plone-frontend-deployment-c872e367",
    },
    "spec": {
      "replicas": 2,
      "selector": {
        "matchLabels": {
          "app": "app-plone-frontend-c8d010d1",
        },
      },
      "template": {
        "metadata": {
          "labels": {
            "app": "app-plone-frontend-c8d010d1",
          },
        },
        "spec": {
          "containers": [
            {
              "env": [
                {
                  "name": "RAZZLE_INTERNAL_API_PATH",
                  "value": "http://app-plone-backend-service-c8efef4c:80",
                },
              ],
              "image": "plone/plone-frontend:latest",
              "imagePullPolicy": "ifNotPresent",
              "name": "frontend-container",
            },
          ],
        },
      },
    },
  },
  {
    "apiVersion": "policy/v1",
    "kind": "PodDisruptionBudget",
    "metadata": {
      "name": "app-plone-frontend-pdb-c82dd14e",
    },
    "spec": {
      "minAvailable": 1,
      "selector": {
        "matchLabels": {
          "app": "app-plone-frontend-c8d010d1",
        },
      },
    },
  },
  {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": {
      "name": "app-plone-frontend-service-c8f392cd",
    },
    "spec": {
      "clusterIP": "None",
      "ports": [
        {
          "port": 80,
          "targetPort": 3000,
        },
      ],
      "selector": {
        "app": "app-plone-frontend-c8d010d1",
      },
      "type": "ClusterIP",
    },
  },
]
`;

exports[`defaults-with-pdps 1`] = `
[
  {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": {
      "name": "app-plone_with_pdbs-backend-deployment-c8d5991a",
    },
    "spec": {
      "replicas": 2,
      "selector": {
        "matchLabels": {
          "app": "app-plone_with_pdbs-backend-c885cd24",
        },
      },
      "template": {
        "metadata": {
          "labels": {
            "app": "app-plone_with_pdbs-backend-c885cd24",
          },
        },
        "spec": {
          "containers": [
            {
              "env": [],
              "image": "plone/plone-backend:latest",
              "imagePullPolicy": "ifNotPresent",
              "name": "backend-container",
            },
          ],
        },
      },
    },
  },
  {
    "apiVersion": "policy/v1",
    "kind": "PodDisruptionBudget",
    "metadata": {
      "name": "app-plone_with_pdbs-backend-pdb-c810ca80",
    },
    "spec": {
      "maxUnavailable": 1,
      "selector": {
        "matchLabels": {
          "app": "app-plone_with_pdbs-backend-c885cd24",
        },
      },
    },
  },
  {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": {
      "name": "app-plone_with_pdbs-backend-service-c819541f",
    },
    "spec": {
      "clusterIP": "None",
      "ports": [
        {
          "port": 80,
          "targetPort": 8080,
        },
      ],
      "selector": {
        "app": "app-plone_with_pdbs-backend-c885cd24",
      },
      "type": "ClusterIP",
    },
  },
  {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": {
      "name": "app-plone_with_pdbs-frontend-deployment-c812b31a",
    },
    "spec": {
      "replicas": 2,
      "selector": {
        "matchLabels": {
          "app": "app-plone_with_pdbs-frontend-c850bc83",
        },
      },
      "template": {
        "metadata": {
          "labels": {
            "app": "app-plone_with_pdbs-frontend-c850bc83",
          },
        },
        "spec": {
          "containers": [
            {
              "env": [
                {
                  "name": "RAZZLE_INTERNAL_API_PATH",
                  "value": "http://app-plone_with_pdbs-backend-service-c819541f:80",
                },
              ],
              "image": "plone/plone-frontend:latest",
              "imagePullPolicy": "ifNotPresent",
              "name": "frontend-container",
            },
          ],
        },
      },
    },
  },
  {
    "apiVersion": "policy/v1",
    "kind": "PodDisruptionBudget",
    "metadata": {
      "name": "app-plone_with_pdbs-frontend-pdb-c8ebc6c1",
    },
    "spec": {
      "minAvailable": 2,
      "selector": {
        "matchLabels": {
          "app": "app-plone_with_pdbs-frontend-c850bc83",
        },
      },
    },
  },
  {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": {
      "name": "app-plone_with_pdbs-frontend-service-c8999de8",
    },
    "spec": {
      "clusterIP": "None",
      "ports": [
        {
          "port": 80,
          "targetPort": 3000,
        },
      ],
      "selector": {
        "app": "app-plone_with_pdbs-frontend-c850bc83",
      },
      "type": "ClusterIP",
    },
  },
]
`;
